src/apps/parser.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2017, [Ribose Inc](https://www.ribose.com).
+ * All rights reserved.
+ *
+ * This code is originally derived from software contributed by
+ * Ribose Inc (https://www.ribose.com).
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** \file
+ * Parser demo app
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+
+#include <repgp/repgp.h>
+
+#include <rnp/rnp.h>             // for rnp_t, rnp_ctx_t et. all
+#include <rekey/rnp_key_store.h> // for keystore stuff
+
+#define RING "keyring"
+
+static bool
+configure_rnp(rnp_t *rnp)
+{
+    rnp_params_t params = {0};
+    bool         res = true;
+
+    // IO
+    pgp_io_t pgpio = {.errs = stderr, .res = stdout, .outs = stdout};
+    rnp->io = malloc(sizeof(pgp_io_t));
+    memcpy(rnp->io, &pgpio, sizeof(pgp_io_t));
+
+    // RNP
+    params.pubpath = strdup(RING "pubring.gpg");
+    params.secpath = strdup(RING "secring.gpg");
```

Comment:
strdup may return `NULL`, I can't belive to this here, but it still may return

Reply:
Fixed